File: Sim/icarus_sim/founding.py
"""Seeded parent-race founding rounds; geographic distances are physical metres."""
import random
from .terrain_tectonics import child_seed


def found_cities(seed, parents, owners, quotas, candidates, scores, distance, spacing, world_reach, survivors=(), forbidden=(), years_per_round=250, origin_min_separation_degrees=90, start_year=0, diaspora_interval_years=1000, diaspora_bonus_per_parent=1, food_allowances=None, city_limit=24, diaspora_bonus_used=(), used_civilizations=(), magic_enabled=True):
    sites=[dict(s) for s in survivors];blocked=set(forbidden);events=[]
    quotas=dict(quotas);bonus_used=set(diaspora_bonus_used);ever_used=set(used_civilizations)|{s['population_profile'] for s in sites}
    origin_gap=world_reach*origin_min_separation_degrees/180
    year_offset=max(start_year,max((s.get('founding_year',0) or 0 for s in sites),default=0)+(years_per_round if sites else 0))
    blocked_parents=set()
    target=sum(quotas.values());base=max(spacing*2,world_reach/32,1)
    def available(race):
        counts={key:sum(s['population_profile']==key for s in sites) for key in owners}
        return [(key,node) for key in owners if owners[key]==race and counts[key]<quotas[key]
                for node in candidates[key] if node not in blocked and all(node!=s['node'] and distance(node,s['node'])>=spacing for s in sites)]
    def diaspora_candidates():
        if food_allowances is None or len(sites)>=city_limit:return {}
        options={}
        for key,race in owners.items():
            if key in ever_used or food_allowances.get(key,0)<40:continue
            bonus=quotas[key]==0
            if bonus and (not diaspora_bonus_per_parent or race in bonus_used or target>=city_limit):continue
            if len(sites)>=target and not bonus:continue
            members=[s for s in sites if s['parent_race_id']==race]
            pool=[node for node in candidates[key] if node not in blocked and all(node!=s['node'] and distance(node,s['node'])>=spacing for s in sites)]
            if not members:pool=[node for node in pool if all(distance(node,s['node'])+1e-8>=origin_gap for s in sites if s.get('founding_capital',False))]
            if pool:options[key]=pool
        return options
    stop='round_limit_reached'
    for turn in range(1,129):
        radius=min(world_reach,base*1.6**min(turn-1,32));possible=False
        # Rotate initiative so one parent does not always claim contested sites first.
        races=list(parents);offset=(turn-1)%len(races) if races else 0;races=races[offset:]+races[:offset]
        for race in races:
            if len(sites)>=target:break
            if race in blocked_parents:continue
            members=[s for s in sites if s['parent_race_id']==race]
            pool=available(race)
            pct=parents[race]['founding_participation_percent']
            if not pool or ((turn>1 or members) and pct==0):continue
            if not members:
                origins=[s for s in sites if s.get('founding_capital',False)]
                pool=[pair for pair in pool if all(distance(pair[1],s['node'])+1e-8>=origin_gap for s in origins)]
                if not pool:
                    blocked_parents.add(race)
                    events.append(dict(turn=turn,parent_race_id=race,status='no_separated_origin',radius_m=world_reach,required_separation_degrees=origin_min_separation_degrees))
                    continue
            possible=True
            draw=random.Random(child_seed(seed,f'founding-{turn}-{race}')).random()*100
            if (turn>1 or members) and draw>=pct:
                events.append(dict(turn=turn,parent_race_id=race,status='did_not_participate',draw_percent=draw,radius_m=radius));continue
            unused=[pair for pair in pool if not any(s['population_profile']==pair[0] for s in sites)]
            if unused:pool=unused
            if members:pool=[pair for pair in pool if min(distance(pair[1],s['node']) for s in members)<=radius]
            if not pool:
                events.append(dict(turn=turn,parent_race_id=race,status='search_radius_too_small',radius_m=radius));continue
            key,node=min(pool,key=lambda pair:(-scores[pair[0]][pair[1]],pair[0],pair[1]))
            record=dict(node=node,population_profile=key,parent_race_id=race,founding_turn=turn,founding_capital=not members)
            source=min(members,key=lambda s:(distance(node,s['node']),s['node'])) if members else None
            record.update(founding_year=year_offset+(turn-1)*years_per_round,
                          migration_source_node=source['node'] if source else None,
                          source_civilization_id=source['population_profile'] if source else None,
                          migration_distance_m=distance(node,source['node']) if source else 0,
                          cultural_branch=bool(source and source['population_profile']!=key))
            sites.append(record);ever_used.add(key)
            events.append(dict(record,status='founded',turn=turn,radius_m=world_reach if not members else radius,suitability=scores[key][node]))
        year=year_offset+(turn-1)*years_per_round
        options=diaspora_candidates()
        if options and year>0 and year%diaspora_interval_years==0:
            rng=random.Random(child_seed(seed,f'diaspora-{year}'))
            key=rng.choice(sorted(options));race=owners[key]
            node=min(options[key],key=lambda n:(-scores[key][n],n))
            members=[s for s in sites if s['parent_race_id']==race]
            source=min(members,key=lambda s:(distance(node,s['node']),s['node'])) if members else None
            bonus=quotas[key]==0
            if bonus:quotas[key]=1;target+=1;bonus_used.add(race)
            reasons=['exile','religious_schism','expedition']+(['magical_displacement'] if magic_enabled else [])
            record=dict(node=node,population_profile=key,parent_race_id=race,founding_turn=turn,founding_capital=not members,
                        founding_year=year,migration_source_node=source['node'] if source else None,
                        source_civilization_id=source['population_profile'] if source else None,
                        migration_distance_m=distance(node,source['node']) if source else 0,cultural_branch=bool(source),
                        diaspora=True,diaspora_reason=rng.choice(reasons),diaspora_bonus=bonus)
            sites.append(record);ever_used.add(key)
            events.append(dict(record,status='diaspora_founded',turn=turn,radius_m=world_reach,suitability=scores[key][node]))
            options=diaspora_candidates()
        if options:possible=True
        if len(sites)>=target and not options:stop='capacity_target_reached';break
        if not possible:stop='no_eligible_expansion';break
    return sites,dict(version=3,effective_quotas=quotas,diaspora_interval_years=diaspora_interval_years,diaspora_bonus_used=sorted(bonus_used),used_civilizations=sorted(ever_used),start_year=year_offset,end_year=year_offset+(turn-1)*years_per_round,years_per_round=years_per_round,elapsed_years=(turn-1)*years_per_round,origin_min_separation_degrees=origin_min_separation_degrees,turns=turn,target_cities=target,placed_cities=len(sites),stop_reason=stop,events=events,
                      definition='Sufficient population means capacity-derived city quotas, plus at most one food-supported diaspora bonus per parent, subject to safe habitat, spacing and a 24-city lab ceiling; 128-round safety limit.')

File: Core/founding.cpp
#include "founding.hpp"
#include "pyrandom.hpp"
#include <algorithm>
#include <cmath>
#include <set>
namespace fantasy_world_generator {
namespace {
struct Candidate {std::string civilization;std::size_t node=0;};
}
FoundingResult found_cities(std::uint32_t seed,const WorldConfig& cfg,double radius,const SphereGrid& grid,
                            const Catalogues& catalogues,const SettlementFields& fields,
                            const FoundingContext& context) {
    const std::vector<std::string>& civilizations=catalogues.civilization_ids();
    const FoundingRules& rules=catalogues.founding_rules();
    const double spacing=cfg.settlement_spacing;
    const double world_reach=pi*radius;
    std::vector<Vec3> vectors;
    vectors.reserve(grid.points.size());
    for(const auto& point:grid.points) vectors.push_back(direction(point.first,point.second,cfg.size));
    auto separation=[&](std::size_t a,std::size_t b) {
        return radius*std::acos(std::max(-1.,std::min(1.,dot(vectors[a],vectors[b]))));
    };
    // Only the parent races that actually own a civilization take turns.
    std::vector<ParentRace> races;
    for(const ParentRace& race:catalogues.parent_races()) {
        for(const std::string& civilization:civilizations)
            if(catalogues.parent_race_id(civilization)==race.id) {races.push_back(race);break;}
    }
    FoundingResult result;
    std::map<std::string,std::int64_t> quotas;
    for(const std::string& civilization:civilizations) {
        const std::int64_t minimum=static_cast<std::int64_t>(
            catalogues.profile(civilization).trait("minimum_founding_residents"));
        const std::int64_t allowance=fields.budget.allowances.at(civilization);
        quotas[civilization]=std::min(fields.budget.quotas.at(civilization),
                                      minimum>0 ? allowance/minimum : 0);
    }
    // The ceiling actually in force, which the budget has already resolved: the
    // ground's packing, the caller's count, or a real zero for an empty world.
    const std::int64_t ceiling=fields.budget.city_limit;
    const double origin_gap=world_reach*rules.origin_min_separation_degrees/180;
    std::int64_t target=0;
    for(const auto& entry:quotas) target+=entry.second;
    const double base=std::max(std::max(spacing*2,world_reach/32),1.);
    // A later age starts from the cities that survived, and never rebuilds on a ruin.
    result.sites=context.survivors;
    const std::set<std::size_t> forbidden(context.forbidden.begin(),context.forbidden.end());
    std::set<std::string> ever_used(context.used_civilizations.begin(),context.used_civilizations.end());
    std::set<std::string> bonus_used(context.bonus_used.begin(),context.bonus_used.end()),blocked_parents;
    for(const FoundedCity& site:result.sites) ever_used.insert(site.population_profile);
    double year_offset=context.start_year;
    if(!result.sites.empty()) {
        double latest=0.;
        for(const FoundedCity& site:result.sites) latest=std::max(latest,site.founding_year);
        year_offset=std::max(year_offset,latest+rules.years_per_round);
    }
    auto spaced=[&](std::size_t node) {
        if(forbidden.count(node)!=0) return false;
        for(const FoundedCity& site:result.sites)
            if(node==site.node || !(separation(node,site.node)>=spacing)) return false;
        return true;
    };
    // Cities are only ever added inside a round, so a candidate that has lost its
    // spacing can never regain it. Keeping a shrinking list of still-viable nodes per
    // people turns a rescan of every candidate against every city, on every turn, into
    // one pass per city. The list keeps candidate order, so the choice is unchanged:
    // on a world with hundreds of cities this is the difference between founding
    // taking seconds and taking milliseconds.
    std::map<std::string,std::vector<std::size_t>> viable;
    for(const std::string& civilization:civilizations) {
        std::vector<std::size_t> nodes;
        for(std::size_t node:fields.species.at(civilization).candidates)
            if(spaced(node)) nodes.push_back(node);
        viable.emplace(civilization,std::move(nodes));
    }
    auto retire=[&](std::size_t placed) {
        for(auto& entry:viable) {
            std::vector<std::size_t>& nodes=entry.second;
            nodes.erase(std::remove_if(nodes.begin(),nodes.end(),[&](std::size_t node) {
                return node==placed || !(separation(node,placed)>=spacing);
            }),nodes.end());
        }
    };
    auto available=[&](const std::string& race) {
        std::vector<Candidate> pool;
        for(const std::string& civilization:civilizations) {
            if(catalogues.parent_race_id(civilization)!=race) continue;
            std::int64_t placed=0;
            for(const FoundedCity& site:result.sites) placed+=site.population_profile==civilization ? 1 : 0;
            if(placed>=quotas.at(civilization)) continue;
            for(std::size_t node:viable.at(civilization)) pool.push_back(Candidate{civilization,node});
        }
        return pool;
    };
    auto best_candidate=[&](const std::vector<Candidate>& pool) {
        const Candidate* best=nullptr;
        for(const Candidate& option:pool) {
            if(best==nullptr) {best=&option;continue;}
            const double score=fields.species.at(option.civilization).suitability[option.node];
            const double incumbent=fields.species.at(best->civilization).suitability[best->node];
            if(score>incumbent) {best=&option;continue;}
            if(score<incumbent) continue;
            if(option.civilization<best->civilization) {best=&option;continue;}
            if(option.civilization==best->civilization && option.node<best->node) best=&option;
        }
        return best;
    };
    auto nearest_member=[&](const std::string& race,std::size_t node,const FoundedCity*& source) {
        source=nullptr;
        double best=0.;
        for(const FoundedCity& site:result.sites) {
            if(site.parent_race_id!=race) continue;
            const double distance=separation(node,site.node);
            if(source==nullptr || distance<best || (distance==best && site.node<source->node)) {
                best=distance;source=&site;
            }
        }
    };
    std::string stop="round_limit_reached";
    std::int64_t turn=1;
    for(;turn<=128;++turn) {
        const double search=std::min(world_reach,base*std::pow(1.6,static_cast<double>(std::min<std::int64_t>(turn-1,32))));
        bool possible=false;
        // Rotate initiative so one parent never claims every contested site first.
        std::vector<ParentRace> order=races;
        if(!order.empty()) {
            const std::size_t offset=static_cast<std::size_t>((turn-1)%static_cast<std::int64_t>(order.size()));
            std::rotate(order.begin(),order.begin()+static_cast<std::ptrdiff_t>(offset),order.end());
        }
        for(const ParentRace& race:order) {
            if(static_cast<std::int64_t>(result.sites.size())>=target) break;
            if(blocked_parents.count(race.id)) continue;
            bool has_members=false;
            for(const FoundedCity& site:result.sites) if(site.parent_race_id==race.id) {has_members=true;break;}
            std::vector<Candidate> pool=available(race.id);
            if(pool.empty() || ((turn>1 || has_members) && race.founding_participation_percent==0)) continue;
            if(!has_members) {
                // A first city has to stand clear of every other parent's origin.
                std::vector<Candidate> separated;
                for(const Candidate& option:pool) {
                    bool clear=true;
                    for(const FoundedCity& site:result.sites) {
                        if(!site.founding_capital) continue;
                        if(!(separation(option.node,site.node)+1e-8>=origin_gap)) {clear=false;break;}
                    }
                    if(clear) separated.push_back(option);
                }
                pool=separated;
                if(pool.empty()) {blocked_parents.insert(race.id);continue;}
            }
            possible=true;
            PyRandom rng(child_seed(seed,"founding-"+std::to_string(turn)+"-"+race.id));
            const double draw=rng.next()*100;
            if((turn>1 || has_members) && draw>=race.founding_participation_percent) continue;
            std::vector<Candidate> unused;
            for(const Candidate& option:pool) {
                bool taken=false;
                for(const FoundedCity& site:result.sites)
                    if(site.population_profile==option.civilization) {taken=true;break;}
                if(!taken) unused.push_back(option);
            }
            if(!unused.empty()) pool=unused;
            if(has_members) {
                std::vector<Candidate> reachable;
                for(const Candidate& option:pool) {
                    double nearest=0.;
                    bool first=true;
                    for(const FoundedCity& site:result.sites) {
                        if(site.parent_race_id!=race.id) continue;
                        const double distance=separation(option.node,site.node);
                        if(first || distance<nearest) {nearest=distance;first=false;}
                    }
                    if(!first && nearest<=search) reachable.push_back(option);
                }
                pool=reachable;
            }
            if(pool.empty()) continue;
            const Candidate* chosen=best_candidate(pool);
            const FoundedCity* source=nullptr;
            if(has_members) nearest_member(race.id,chosen->node,source);
            FoundedCity city;
            city.node=chosen->node;
            city.population_profile=chosen->civilization;
            city.parent_race_id=race.id;
            city.founding_turn=turn;
            city.founding_capital=!has_members;
            city.founding_year=year_offset+static_cast<double>(turn-1)*rules.years_per_round;
            if(source!=nullptr) {
                city.has_source=true;
                city.migration_source_node=source->node;
                city.source_civilization_id=source->population_profile;
                city.migration_distance_m=separation(chosen->node,source->node);
                city.cultural_branch=source->population_profile!=chosen->civilization;
            }
            result.sites.push_back(city);
            retire(city.node);
            ever_used.insert(chosen->civilization);
        }
        const double year=year_offset+static_cast<double>(turn-1)*rules.years_per_round;
        // A diaspora can seat an unused civilization that the quotas never reached.
        auto diaspora_options=[&]() {
            std::map<std::string,std::vector<std::size_t>> options;
            if(static_cast<std::int64_t>(result.sites.size())>=ceiling) return options;
            for(const std::string& civilization:civilizations) {
                const std::int64_t minimum=static_cast<std::int64_t>(
                    catalogues.profile(civilization).trait("minimum_founding_residents"));
                if(ever_used.count(civilization)) continue;
                if(fields.budget.allowances.at(civilization)<minimum) continue;
                const std::string& race=catalogues.parent_race_id(civilization);
                const bool bonus=quotas.at(civilization)==0;
                if(bonus && (rules.diaspora_bonus_per_parent==0 || bonus_used.count(race)
                             || target>=ceiling)) continue;
                if(static_cast<std::int64_t>(result.sites.size())>=target && !bonus) continue;
                bool has_members=false;
                for(const FoundedCity& site:result.sites) if(site.parent_race_id==race) {has_members=true;break;}
                std::vector<std::size_t> pool;
                for(std::size_t node:viable.at(civilization)) {
                    if(!has_members) {
                        bool clear=true;
                        for(const FoundedCity& site:result.sites) {
                            if(!site.founding_capital) continue;
                            if(!(separation(node,site.node)+1e-8>=origin_gap)) {clear=false;break;}
                        }
                        if(!clear) continue;
                    }
                    pool.push_back(node);
                }
                if(!pool.empty()) options.emplace(civilization,pool);
            }
            return options;
        };
        std::map<std::string,std::vector<std::size_t>> options=diaspora_options();
        if(!options.empty() && year>0 && std::fmod(year,rules.diaspora_interval_years)==0) {
            PyRandom rng(child_seed(seed,"diaspora-"+std::to_string(static_cast<std::int64_t>(year))));
            std::vector<std::string> keys;
            for(const auto& entry:options) keys.push_back(entry.first);
            const std::string civilization=keys[rng.choice(keys.size())];
            const std::string& race=catalogues.parent_race_id(civilization);
            std::size_t node=options.at(civilization).front();
            for(std::size_t option:options.at(civilization)) {
                const double score=fields.species.at(civilization).suitability[option];
                const double incumbent=fields.species.at(civilization).suitability[node];
                if(score>incumbent || (score==incumbent && option<node)) node=option;
            }
            const FoundedCity* source=nullptr;
            nearest_member(race,node,source);
            const bool bonus=quotas.at(civilization)==0;
            if(bonus) {quotas.at(civilization)=1;target+=1;bonus_used.insert(race);}
            FoundedCity city;
            city.node=node;
            city.population_profile=civilization;
            city.parent_race_id=race;
            city.founding_turn=turn;
            city.founding_capital=source==nullptr;
            city.founding_year=year;
            city.diaspora=true;
            city.diaspora_bonus=bonus;
            if(source!=nullptr) {
                city.has_source=true;
                city.migration_source_node=source->node;
                city.source_civilization_id=source->population_profile;
                city.migration_distance_m=separation(node,source->node);
                city.cultural_branch=true;
            }
            static const char* reasons[4]={"exile","religious_schism","expedition","magical_displacement"};
            const std::size_t choices=cfg.magic_enabled ? 4 : 3;
            city.diaspora_reason=reasons[rng.choice(choices)];
            result.sites.push_back(city);
            retire(city.node);
            ever_used.insert(civilization);
            options=diaspora_options();
        }
        if(!options.empty()) possible=true;
        if(static_cast<std::int64_t>(result.sites.size())>=target && options.empty()) {
            stop="capacity_target_reached";break;
        }
        if(!possible) {stop="no_eligible_expansion";break;}
    }
    result.turns=std::min<std::int64_t>(turn,128);
    result.target_cities=target;
    result.effective_quotas=quotas;
    result.stop_reason=stop;
    result.diaspora_bonus_used.assign(bonus_used.begin(),bonus_used.end());
    result.used_civilizations.assign(ever_used.begin(),ever_used.end());
    result.start_year=year_offset;
    result.end_year=year_offset+static_cast<double>(result.turns-1)*rules.years_per_round;
    return result;
}
}
